### PR TITLE
allow INST to be used dynamically

### PR DESCRIPTION
**Pull request by stylewarning**
This commit doesn't change the how any existing functionality works,
but does permit new functionality.

Currently, INST & co may only be used lexically within a WITH-INST,
and by extension, within DEFINE-COMPILER. If INST is used outside of
either of those contexts, there will be a _compile-time_ error it can
be detected, but definitely a run-time error.

This commit allows INST to be used anywhere within the dynamic extent
of WITH-INST (and DEFINE-COMPILER). This means it's now possible to
write functions like:

    (defun inst-hadamard (N)
      (dotimes (q N)
        (inst "H" () q)))

There are two downsides:

- Compile-time errors cannot be detected since any INST call may
      possible be valid.

- There may be a small performance penalty using dynamic
      features. However, what we lost in making it dynamic might be
      gained back by much smaller code. (INST and INST* aren't
      duplicated in every definition.)

### CODE REVIEW

**Comment by stylewarning** [comment]
(May be of interest to @ecpeterson.)

File: src/define-compiler.lisp [hunk 1]
```diff
@@ -903,85 +903,94 @@ N.B.: This routine is somewhat fragile, and highly creative compiler authors wil
...
+  (throw +inst-tag+ (if retval-p retval (%collected-insts))))
+
+(defun inst (&rest xs)
+  "Emit the instructions XS in the current compiler context.
+
+INST can only be called within the dynamic extent of a compiler body."
+  (declare (dynamic-extent xs))
+  (unless (boundp '*inst-collection*)
+    (error "INST called outside of a WITH-INST context."))
+  ;; INST is a no-op if something is provided.
```
Comment (ecpeterson) [inline src/define-compiler.lisp]: if nothing**

**Comment by stylewarning** [comment]
For N = 1 samples, we are actually 10 seconds _faster_.

before:
```
Evaluation took:
  462.272 seconds of real time
  463.079237 seconds of total run time (460.666083 user, 2.413154 system)
  [ Run times consist of 3.633 seconds GC time, and 459.447 seconds non-GC time. ]
  100.17% CPU
  92,424 forms interpreted
  13,567 lambdas converted
  1,109,454,218,200 processor cycles
  323,737,789,280 bytes consed
```

after:
```
  451.827 seconds of real time
  452.877278 seconds of total run time (450.658180 user, 2.219098 system)
  [ Run times consist of 3.461 seconds GC time, and 449.417 seconds non-GC time. ]
  100.23% CPU
  92,424 forms interpreted
  4,796 lambdas converted
  1,084,334,105,504 processor cycles
  320,138,089,840 bytes consed
```

**Comment by stylewarning** [comment]
I still request @appleby looks at it.

File: src/define-compiler.lisp [hunk 2]
```diff
@@ -903,85 +903,94 @@ N.B.: This routine is somewhat fragile, and highly creative compiler authors wil
...
+This macro is used for DEFINE-COMPILER."
+  `(let ((*inst-collection* (%new-inst-collection)))
+     (catch ',+inst-tag+
+       ,@body
+       (%collected-insts))))            ; FINISH-COMPILER would also be OK.
 
-FINISH-COMPILER is a local macro usable within a compiler body."
-  (error "FINISH-COMPILER can only be used in the body of a compiler."))
+(defun finish-compiler (&optional (retval nil retval-p))
+  "Finish compiling, by default returning a list of all instructed emitted with INST.
```
Comment (appleby) [inline src/define-compiler.lisp]: instructed -> instructions

Comment (stylewarning) [reply src/define-compiler.lisp:948]: too lazy to fix

**Pull request merged**